harmony_py/auth.py

```diff
@@ -0,0 +1,146 @@
+from typing import cast, Optional, Tuple, Union
+
+from .config import Config
+
+
+cfg = Config()
+
+
+def _is_edl_hostname(hostname: str) -> bool:
+    """
+    Determine if a hostname matches an EDL hostname.
+
+    Parameters:
+        hostname (str): A fully-qualified domain name (FQDN).
+
+    Returns:
+        (boolean): True if the hostname is an EDL hostname, else False.
+    """
+    edl_hostname_pattern = r'.*urs\.earthdata\.nasa\.gov$'
+    return re.fullmatch(edl_hostname_pattern, hostname, flags=re.IGNORECASE) is not None
+
+
+class MalformedCredentials(Exception):
+    pass
+
+
+class BadAuthentication(Exception):
+    pass
+
+
+class SessionWithHeaderRedirection(Session):
+    """Modify Authorization headers in accordance with Earthdata Login (EDL) common usage.
+
+    Example:
+        session = SessionWithHeaderRedirection(username, password)
+
+    Parameters:
+        auth (Tuple(str, str)): A tuple of the format ('edl_username', 'edl_password')
+    """
+
+    def __init__(self, auth: Optional[Union[Tuple[str, str], None]] = None) -> None:
+        super().__init__()
+        if auth:
+            self.auth = auth
+        else:
+            self.auth = None
+
+    def rebuild_auth(self, prepared_request: PreparedRequest, response: Response) -> None:
+        """
+        Override Session.rebuild_auth. Strips the Authorization header if neither
+        original URL nor redirected URL belong to an Earthdata Login (EDL) host. Also
+        allows the default requests behavior of searching for relevant .netrc
+        credentials if and only if a username and password weren't provided during
+        object instantiation.
+
+        Parameters:
+            prepared_request (:obj:`PreparedRequest`): Object for the redirection
+            destination.
+            response (:obj:`PreparedRequest`): Object for the where we just came from.
+
+        Returns:
+            (boolean): True if the hostname is an EDL hostname, else False.
+        """
+
+        headers = prepared_request.headers
+        redirect_hostname = cast(str, urlparse(prepared_request.url).hostname)
+        original_hostname = cast(str, urlparse(response.request.url).hostname)
+
+        if 'Authorization' in headers \
+                and (original_hostname != redirect_hostname) \
+                and not _is_edl_hostname(redirect_hostname):
+            del headers['Authorization']
+
+        if self.auth is None:
+            # .netrc might have more auth for us on our new host.
+            new_auth = get_netrc_auth(prepared_request.url) if self.trust_env else None
+            if new_auth is not None:
+                prepared_request.prepare_auth(new_auth)
+
+        return
+
+
+def authenticate(*, auth: Optional[Union[Tuple[str, str], None]] = None,
+                 validate: Optional[bool] = True) -> Session:
+    """
+    Create a requests-futures session for authenticated HTTP calls after optionally validating
```

Comment:
What's the expected use of this session?

I was thinking (as alluded to in prior comments) that it'd be something like:
```
harmony_client = harmony.authenticate(...)
harmony_client.subset(...)
```
but is it actually
```
session = harmony.authenticate(...)
harmony.subset(session, ...)
```

Reply:
That makes sense - it should be more of a 'client' (your first example).